File: ios/TimerStore.swift
import Foundation

/**
 * TimerStore
 *
 * Persists TimerData to UserDefaults so timers survive app kills.
 * Uses JSON encoding/decoding via Codable.
 */
final class TimerStore {

  private static let suite = "com.rick427.backgroundtimer"
  private static let keyPrefix = "timer_"

  private let defaults: UserDefaults

  init() {
    self.defaults = UserDefaults(suiteName: Self.suite) ?? .standard
  }

  // ── Write ──────────────────────────────────────────────────────────────────

  func save(_ timer: TimerData) {
    guard timer.persist else { return }
    guard let data = try? JSONEncoder().encode(timer) else { return }
    defaults.set(data, forKey: key(for: timer.id))
  }

  func delete(id: String) {
    defaults.removeObject(forKey: key(for: id))
  }

  func clear() {
    defaults.dictionaryRepresentation()
      .keys
      .filter { $0.hasPrefix(Self.keyPrefix) }
      .forEach { defaults.removeObject(forKey: $0) }
  }

  // ── Read ───────────────────────────────────────────────────────────────────

  func load(id: String) -> TimerData? {
    guard let data = defaults.data(forKey: key(for: id)) else { return nil }
    return try? JSONDecoder().decode(TimerData.self, from: data)
  }

  func loadAll() -> [TimerData] {
    return defaults.dictionaryRepresentation()
      .filter { $0.key.hasPrefix(Self.keyPrefix) }
      .compactMap { (_, value) -> TimerData? in
        guard let data = value as? Data else { return nil }
        return try? JSONDecoder().decode(TimerData.self, from: data)
      }
  }

  // ── Helpers ────────────────────────────────────────────────────────────────

  private func key(for id: String) -> String {
    "\(Self.keyPrefix)\(id)"
  }
}

File: android/src/main/java/com/rick427/backgroundtimer/TimerStore.kt
package com.rick427.backgroundtimer

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONObject

/**
 * Persists timer state to SharedPreferences so it survives app kills.
 * Key schema: "timer_<id>" → JSON string of TimerData
 */
class TimerStore(context: Context) {

  private val prefs: SharedPreferences =
    context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

  // ── Write ─────────────────────────────────────────────────────────────────

  fun save(timer: TimerData) {
    if (!timer.persist) return
    prefs.edit()
      .putString(timerKey(timer.id), timer.toJson().toString())
      .apply()
  }

  fun delete(id: String) {
    prefs.edit().remove(timerKey(id)).apply()
  }

  fun clear() {
    prefs.edit()
      .apply { prefs.all.keys.filter { it.startsWith(KEY_PREFIX) }.forEach { remove(it) } }
      .apply()
  }

  // ── Read ──────────────────────────────────────────────────────────────────

  fun load(id: String): TimerData? {
    val raw = prefs.getString(timerKey(id), null) ?: return null
    return runCatching { TimerData.fromJson(JSONObject(raw)) }.getOrNull()
  }

  fun loadAll(): List<TimerData> {
    return prefs.all.entries
      .filter { it.key.startsWith(KEY_PREFIX) }
      .mapNotNull { entry ->
        runCatching { TimerData.fromJson(JSONObject(entry.value as String)) }.getOrNull()
      }
  }

  private fun timerKey(id: String) = "$KEY_PREFIX$id"

  companion object {
    private const val PREFS_NAME = "RNBackgroundTimer"
    private const val KEY_PREFIX = "timer_"
  }
}
